engine/class_modules/paladin/sc_paladin.cpp

```diff
@@ -3284,39 +3284,67 @@ std::unique_ptr<expr_t> paladin_t::create_expression( util::string_view name_str
     cooldown_t* j_cd;
     cooldown_t* how_cd;
     cooldown_t* wake_cd;
+    cooldown_t* hs_cd;
+    cooldown_t* at_cd;
 
     time_to_hpg_expr_t( util::string_view n, paladin_t& p )
       : paladin_expr_t( n, p ),
         cs_cd( p.get_cooldown( "crusader_strike" ) ),
         boj_cd( p.get_cooldown( "blade_of_justice" ) ),
         j_cd( p.get_cooldown( "judgment" ) ),
         how_cd( p.get_cooldown( "hammer_of_wrath" ) ),
-        wake_cd( p.get_cooldown( "wake_of_ashes" ) )
+        wake_cd( p.get_cooldown( "wake_of_ashes" ) ),
+        hs_cd( p.get_cooldown( "holy_shock" ) ),
+        at_cd( p.get_cooldown( "arcane_torrent" ) )
     {
     }
 
     double evaluate() override
     {
-      if ( paladin.specialization() != PALADIN_RETRIBUTION )
+      if ( paladin.specialization() == PALADIN_PROTECTION )
       {
-        paladin.sim->errorf( "\"time_to_hpg\" only supported for Retribution" );
+        paladin.sim->errorf( "\"time_to_hpg\" not supported for Protection" );
         return 0;
       }
       timespan_t gcd_ready = paladin.gcd_ready - paladin.sim->current_time();
       gcd_ready            = std::max( gcd_ready, 0_ms );
 
       timespan_t shortest_hpg_time = cs_cd->remains();
 
-      if ( boj_cd->remains() < shortest_hpg_time )
-        shortest_hpg_time = boj_cd->remains();
+      // Blood Elf
+      if ( paladin.race == RACE_BLOOD_ELF )
+      {
+        if ( at_cd->remains() < shortest_hpg_time )
+          shortest_hpg_time = at_cd->remains();
+      }
+
+      // Retribution
+      if ( paladin.specialization() == PALADIN_RETRIBUTION )
+      {
+        if ( boj_cd->remains() < shortest_hpg_time )
+          shortest_hpg_time = boj_cd->remains();
+
+        if ( wake_cd->remains() < shortest_hpg_time )
+          shortest_hpg_time = wake_cd->remains();
+
+        if ( j_cd->remains() < shortest_hpg_time )
+          shortest_hpg_time = j_cd->remains();
```

Comment:
@OrcinusDrake @Collisionc for some reason judgment was not included here, even though the CD was above and it does generate holy power for ret, so ive added that back in here. Please review and let me know if i should remove this or if it was a bug

Reply:
Before:
![before-ret](https://user-images.githubusercontent.com/10604059/132264593-f428c6aa-f363-4948-9bf1-1c390e370b95.png)


After:
![after-ret](https://user-images.githubusercontent.com/10604059/132264603-13db67a4-1e62-4168-9cb2-2e590eb593b1.png)